arch/x86/core/intel64/locore.S

```diff
@@ -0,0 +1,586 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <arch/x86/multiboot.h>
+#include <sys/util.h>
+#include <arch/x86/msr.h>
+#include <kernel_arch_data.h>
+#include <offsets_short.h>
+#include <drivers/interrupt_controller/loapic.h>
+#include <arch/cpu.h>
+
+.section .locore,"ax"
+.code32
+
+.globl __start
+__start:
+
+	/*
+	 * multiboot starts us here in 32-bit flat mode with interrupts and
+	 * paging disabled. first, check to see if EAX has a valid multiboot
+	 * magic in it, and if not, zero EBX so we know it's bogus later.
+	 * Remember not to clobber EBX until we've copied the boot info struct!
+	 */
+
+#ifdef CONFIG_X86_MULTIBOOT_INFO
+	cmpl $X86_MULTIBOOT_EAX_MAGIC, %eax
+	je 1f
+	xorl %ebx, %ebx
+1:
+#endif
+
+	/* FIXME: ...disabling the PICs is common to IA32 and Intel64... */
+
+#ifdef CONFIG_PIC_DISABLE
+	movb $0xff, %al
+	outb %al, $0x21
+	outb %al, $0xA1
+#endif
+
+	/* next, switch to our own GDT/IDT and stack. */
+
+	lgdt gdt48
+	lidt idt48
+	jmpl $X86_KERNEL_CS_32, $1f
+1:	movw $X86_KERNEL_DS_32, %ax
+	movw %ax, %ds
+	movw %ax, %ss
+	movw %ax, %es
+	movw %ax, %fs
+	movw %ax, %gs
+
+	movl $(exception_stack + CONFIG_EXCEPTION_STACK_SIZE), %esp
+
+	/* transition to long mode. along the way, we enable SSE. */
+
+	movl %cr4, %eax			/* enable PAE and SSE */
+	orl $(CR4_PAE | CR4_OSFXSR), %eax
+	movl %eax, %cr4
+	clts
+
+	movl $pml4, %eax		/* load page base */
+	movl %eax, %cr3
+
+	movl $X86_EFER_MSR, %ecx	/* enable long mode */
+	rdmsr
+	orl $X86_EFER_MSR_LME, %eax
+	wrmsr
+
+	movl %cr0, %eax			/* enable paging */
+	orl $CR0_PG, %eax
+	movl %eax, %cr0
+
+	/* jump into long mode, reload the segment registers (again). */
+
+	jmpl $X86_KERNEL_CS_64, $1f
+.code64
+1:	movl $X86_KERNEL_DS_64, %eax
+	movw %ax, %ds
+	movw %ax, %es
+	movw %ax, %ss
+	movw %ax, %fs
+	movw %ax, %gs
+
+	movl $X86_KERNEL_TSS, %eax
+	ltr %ax
+
+	cld
+	xorl %eax, %eax
+	movq $__bss_start, %rdi
+	movq $__bss_num_qwords, %rcx
+	rep stosq
+
+#ifdef CONFIG_X86_MULTIBOOT_INFO
+	/* If EBX isn't 0, then we have a valid multiboot info struct. */
+
+	testl %ebx, %ebx
+	jz 1f
+	movl %ebx, %esi
+
+	movl $__X86_MULTIBOOT_INFO_SIZEOF, %ecx
+	movq $x86_multiboot_info, %rdi
+	rep
+	movsb
+1:
+#endif /* CONFIG_X86_MULTIBOOT_INFO */
+
+	/*
+	 * set up SSE in case something uses the floating-point unit during
+	 * early initialization (either directly, or if GCC gets clever)
+	 */
+
+	xorl %edi, %edi
+	call x86_sse_init
+
+	/* don't replace CALL with JMP; honor the ABI stack alignment! */
+
```

Comment:
> Somewhere in here I think we'll need a block for CONFIG_INIT_STACKS for the IRQ stack buffers, which is an optional feature to pre-fill all stack buffers with 0xAA

No problem.  Will add.

> Could also just move the definition of ist to asm code and .fill with 0xAA

Unfortunately if we do that, because there are so many stacks and they're huge, we'll bloat the load image. Since this is a configurable option, I'll take the start-up delay by filling at runtime.

Reply:
Fixed.